### PR TITLE
Nlx unique chan id fix

### PR DESCRIPTION
**Pull request by sin-mike**
@samuelgarcia 

This PR includes a fix for non-unique chan_id when there are several files for the same AD channel

see  #590,  #589

Now I can read mixed records without splitting folders, like this
image

These two channels are reading from one AD channel, but with different preprocessings:
image

### CODE REVIEW

**Comment by sin-mike** [comment]
Hi, @samuelgarcia, any updates on this PR?

**Comment by sin-mike** [comment]
> 
> 
> @sin-mike
> Thank you very much for this. This is cool to fix.
> I was am a bit busy and still be until wensday.
> I will make comment soon.
> 
> For short, I am a bit discomfort with this uuid, this make the logic harder to understand.
> In my mind the double key (name, chan_id) was already unique for the dict.

@samuelgarcia Indeed, but here I was worried about the consistency, as, in the header, you're using just integer ids. And that using tuple ids could easily break something further. As I discovered, the header is a key structure for the data IO operations, e.g. to restrict IOs to certain channels, I just modified the header and removed rows partially. Therefore I used just integer IDs.

**Comment by samuelgarcia** [comment]
@sin-mike I am back.
thank you for this new comit but I had in mind the posite solution : remove this generated uid.
I think we can go wihtout it. 

In short in the header['signal_channels'] the chan_id must totally consistent with what the user known about the file. So we must put the real chan_id even it is not unique.
Some IOs do not have unique id or unique names for sig channels. it is not problem because we can always  ```acces then read_signal_chunk(channel_indexes=[...])```. In your special case ```read_signal_chunk(channel_names=[...])``` and ```read_signal_chunk(channel_ids=[...])``` will raise an error because it is tested at the BaseRawIO class level.

Other remark: why did you change the id for spike channel ?

I will make comment in the code.

If we use a generated id it should be internal but in our case I think it is useless.

I feel really sorry to add more work.

File: neo/rawio/neuralynxrawio.py [hunk 1]
```diff
@@ -65,31 +65,41 @@ def _parse_header(self):
         unit_channels = []
         event_channels = []
 
-        self.ncs_filenames = OrderedDict()  # chan_id: filename
-        self.nse_ntt_filenames = OrderedDict()  # chan_id: filename
+        self.ncs_filenames = OrderedDict()  # uid: filename
```
Comment (samuelgarcia) [inline neo/rawio/neuralynxrawio.py]: We can do this with a tupple for the key:
self.ncs_filenames = OrderedDict()  # (chan_name, chan_id): filename

File: neo/rawio/neuralynxrawio.py [hunk 2]
```diff
@@ -65,31 +65,41 @@ def _parse_header(self):
...
+        self.nse_ntt_filenames = OrderedDict()  # uid: filename
```
Comment (samuelgarcia) [inline neo/rawio/neuralynxrawio.py]: I prefer also :
 # (chan_name, chan_id): filename

File: neo/rawio/neuralynxrawio.py [hunk 3]
```diff
@@ -65,31 +65,41 @@ def _parse_header(self):
...
         self.nev_filenames = OrderedDict()  # chan_id: filename
+        self.uid2chid = OrderedDict() # uid : (file_enum, chan_enum, chan_id)
```
Comment (samuelgarcia) [inline neo/rawio/neuralynxrawio.py]: I prefer also :
 (chan_name, chan_id): filename

File: neo/rawio/neuralynxrawio.py [hunk 4]
```diff
@@ -65,31 +65,41 @@ def _parse_header(self):
...
         # And construct channels headers
         signal_annotations = []
         unit_annotations = []
         event_annotations = []
 
-        for filename in sorted(os.listdir(self.dirname)):
+        # filter all filenames to match IO extensions,
+        # so that fid indicates only the necessary files
+        all_filenames = filter(lambda f: os.path.splitext(f)[1][1:] in self.extensions, 
+                               sorted(os.listdir(self.dirname)))
```
Comment (samuelgarcia) [inline neo/rawio/neuralynxrawio.py]: Super good.

File: neo/rawio/neuralynxrawio.py [hunk 5]
```diff
@@ -65,31 +65,41 @@ def _parse_header(self):
...
+
+        # I'm generating a unique id for a channel.
+        # chan_id is not a unique id, as several
+        # CSCs could reference the same chan_id
+        # // Mike
+        gen_uid = lambda x,y,z: x + 1000*y
+
+        for fid, filename in enumerate(all_filenames):
```
Comment (samuelgarcia) [inline neo/rawio/neuralynxrawio.py]: I am quite unhappy with this id.

File: neo/rawio/neuralynxrawio.py [hunk 6]
```diff
@@ -98,19 +108,27 @@ def _parse_header(self):
...
             chan_ids = info['channel_ids']
 
-            for idx, chan_id in enumerate(chan_ids):
-                chan_name = chan_names[idx]
+            for idx, (chan_id, chan_name) in enumerate(zip(chan_ids, chan_names)):
+                # ^^^ I had never seen cases when idx is non-zero // Mike
+                
+                uid = gen_uid(fid, idx, chan_id)
+                self.uid2chid[uid] = (fid, idx, chan_id)
+
```
Comment (samuelgarcia) [inline neo/rawio/neuralynxrawio.py]: This could be remove so.

File: neo/rawio/neuralynxrawio.py [hunk 7]
```diff
@@ -98,19 +108,27 @@ def _parse_header(self):
...
                     units = 'uV'
                     gain = info['bit_to_microVolt'][idx]
                     if info['input_inverted']:
                         gain *= -1
                     offset = 0.
                     group_id = 0
-                    sig_channels.append((chan_name, chan_id, info['sampling_rate'],
+
+                    ### Note: vvvv: this structure could contain non-unique chan_name, chan_id
+                    sig_channels.append((chan_name, uid, info['sampling_rate'],
```
Comment (samuelgarcia) [inline neo/rawio/neuralynxrawio.py]: Let s go back to
sig_channels.append((chan_name, chan_id, info['sampling_rate'],
so the use can have something that match the real id in the file and the software.

File: neo/rawio/neuralynxrawio.py [hunk 8]
```diff
@@ -98,19 +108,27 @@ def _parse_header(self):
...
                                          'int16', units, gain, offset, group_id))
-                    self.ncs_filenames[chan_id] = filename
+                    ### // Mike
+
+                    self.ncs_filenames[uid] = filename
```
Comment (samuelgarcia) [inline neo/rawio/neuralynxrawio.py]: self.ncs_filenames[( chan_name, chan_id)] = filename

File: neo/rawio/neuralynxrawio.py [hunk 9]
```diff
@@ -133,26 +151,36 @@ def _parse_header(self):
...
                     # a file can contain several unit_id (so several unit channel)
-                    assert chan_id not in self.nse_ntt_filenames, \
-                        'Several nse or ntt files have the same unit_id!!!'
-                    self.nse_ntt_filenames[chan_id] = filename
+
+                    # vvv this assert is not informative anymore
+                    # assert chan_id not in self.nse_ntt_filenames, \
+                    #     'Several nse or ntt files have the same unit_id!!!'
+
+                    self.nse_ntt_filenames[uid] = filename
```
Comment (samuelgarcia) [inline neo/rawio/neuralynxrawio.py]: self.nse_ntt_filenames[( chan_name, chan_id)] = filename

File: neo/rawio/neuralynxrawio.py [hunk 10]
```diff
@@ -133,26 +151,36 @@ def _parse_header(self):
...
-                    data = np.memmap(filename, dtype=dtype, mode='r', offset=HEADER_SIZE)
-                    self._spike_memmap[chan_id] = data
+
+                    if (os.path.getsize(filename)<=HEADER_SIZE):
+                        self._empty_nse_ntt.append(filename)
+                        data = np.zeros((0,), dtype=dtype)
+                    else:
+                        data = np.memmap(filename, dtype=dtype, mode='r', offset=HEADER_SIZE)
+
+                    self._spike_memmap[uid] = data
```
Comment (samuelgarcia) [inline neo/rawio/neuralynxrawio.py]: OK

File: neo/rawio/neuralynxrawio.py [hunk 11]
```diff
@@ -133,26 +151,36 @@ def _parse_header(self):
...
 
                     unit_ids = np.unique(data['unit_id'])
                     for unit_id in unit_ids:
                         # a spike channel for each (chan_id, unit_id)
-                        self.internal_unit_ids.append((chan_id, unit_id))
+                        self.internal_unit_ids.append((uid, unit_id))
```
Comment (samuelgarcia) [inline neo/rawio/neuralynxrawio.py]: This could disappear.

File: neo/rawio/neuralynxrawio.py [hunk 12]
```diff
@@ -133,26 +151,36 @@ def _parse_header(self):
...
 
-                        unit_name = "ch{}#{}".format(chan_id, unit_id)
+                        unit_name = "ch{}#{}".format(uid, unit_id)
```
Comment (samuelgarcia) [inline neo/rawio/neuralynxrawio.py]: Let s go back to unit_name = "ch{}#{}".format(chan_id, unit_id) or to
unit_name = "{}ch{}#{}".format(chan_name, chan_id, unit_id)
so the user can make the link with the sofwtare

**Comment by samuelgarcia** [comment]
I made some comment in the code.

Best.

**Comment by sin-mike** [comment]
@samuelgarcia 

Ok, I tried with tuple ids (chan

but...

it breaks at header info formatting:
`206         sig_channels = np.array(sig_channels, dtype=_signal_channel_dtype)`

The signal channel dtypes come from:
`from .baserawio import (BaseRawIO, _signal_channel_dtype,   _unit_channel_dtype, _event_channel_dtype)`

So, in my version, it was just:
`sig_channels.append((chan_name, uid, info['sampling_rate'], 'int16', units, gain, offset, group_id))`

Then, with tuple ids, I must write: 
`sig_channels.append((chan_name, (chan_name, chan_id), info['sampling_rate'], 'int16', units, gain, offset, group_id))`

with tuples, it breaks the header dtypes, and I tough the header is something built in stone, so I'm not allowed to change its fields.

If I'm allowed to change the header structure, then, no problem, tuple ids will work, otherwise, `id` field must be ints

**Comment by samuelgarcia** [comment]
(chan_name, chan_id) would not be for the header['signal_channels'] but for self.ncs_filenames[ (chan_name, chan_id)]

**Pull request closed without merging**